### PR TITLE
Add no optimizer approach to update the model

### PR DESCRIPTION
**Pull request by BichengYing**
The motivation of this PR is for minimize the memory consumption. This is just a quick dirty approach, will refactor it later.
I have tested with MNIST dataset and make sure the accuracy increase properly. Please double check the memory assumption.

<s>During the validation, I found a bug....... </s>We definitely need more unit test

### CODE REVIEW

File: gradient_estimators/random_gradient_estimator.py [hunk 1]
```diff
@@ -220,10 +240,20 @@ def update_model_given_seed_and_grad(
...
+            lr = optimizer.defaults["lr"]  # Assume only one parameter group with lr.
+            assert self.paramwise_perturb
+            for one_update_seed, one_update_grad_dirs in zip(
+                iteration_seeds, iteration_grad_scalar
+            ):
+                self.sgd_no_optim_update_model(one_update_grad_dirs, one_update_seed, lr)
+            return
+
         # NOTE: this zero_grad operation is critical since it sets the parameter.grad to None
         # which is checked in self.generate_then_put_grad_paramwise
```
Comment (ZidongLiu) [inline gradient_estimators/random_gradient_estimator.py:254]: Feel the comment seems not relevant anymore?

we assign value when i == 0, this should decouple 2 parts of logic.

File: gradient_estimators/random_gradient_estimator.py [hunk 2]
```diff
@@ -237,15 +267,17 @@ def revert_model_given_seed_and_grad(
...
         try:
             assert isinstance(optimizer, torch.optim.SGD) and optimizer.defaults["momentum"] == 0
         except AssertionError:
             raise Exception("Revert only supports SGD without momentum")
 
         lr, weight_decay = optimizer.defaults["lr"], optimizer.defaults["weight_decay"]
-        optimizer.zero_grad()
         for one_update_seed, one_update_grad_dirs in zip(iteration_seeds, iteration_grad_scalar):
+            optimizer.zero_grad()
             if self.paramwise_perturb:
```
Comment (ZidongLiu) [inline gradient_estimators/random_gradient_estimator.py:281]: Was previous code wrong? Or this is just for memory improvement

**Review by ZidongLiu — COMMENTED** [review]
Code LGTM, but got some questions.

Will do some memory profiling.

File: gradient_estimators/random_gradient_estimator.py (hunk 1 above)
Comment (ZidongLiu) [reply gradient_estimators/random_gradient_estimator.py:254]: like even we do not zero_grad, this code should still work

Comment (BichengYing) [reply gradient_estimators/random_gradient_estimator.py]: right, delete that comment

File: gradient_estimators/random_gradient_estimator.py (hunk 2 above)
Comment (BichengYing) [reply gradient_estimators/random_gradient_estimator.py:279]: after second thoughts, I feel it is better to delete them and add a comment

**Comment by ZidongLiu** [comment]
Memory test is good, 1 client set up, 125m, float32. Model size 500mb. Largest parameter size 147mb. 

server device memory usage: 650mb ~= 500 + 147
image

client device memory usage: 1150mb ~= 2 * 500 + 147
image

**Review by ZidongLiu — APPROVED** [review]
Great work! The memory usage saved is significant. 

Just a last piece, I feel we need some more config options change to unblock @RogerRogerUSC from running experiments
